### PR TITLE
Add Error Capturing Mechanism ref #260

### PR DESCRIPTION
**Pull request by chainsawriot**
This is the error capturing mechanism implemented by @TimBMK 

TODOS

- [x] Add tests
- [x] Port it back to `hydrate_tweets`

### CODE REVIEW

File: tests/testthat/test-errorcapture.R [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+    for (i in seq_len(3)) {
+      index <- c(1, 21, 41)[i]
+      index_end <- index+19
+      x <- fff[index:index_end]
+      params[["ids"]] <- paste0(x, collapse = ",")
+      expect_error(new_rows <- get_tweets(params = params, endpoint_url = endpoint_url, n = Inf, file = NULL, bearer_token = get_bearer(), export_query = FALSE, data_path = emptydir, bind_tweets = TRUE, verbose = FALSE, errors = TRUE), NA)
+      Sys.sleep(0.5)
+      expect_equal(length(list.files(emptydir)), i * 3)
+      expect_equal(length(list.files(emptydir, "^errors")), i)
```
Comment (chainsawriot) [inline tests/testthat/test-errorcapture.R:54]: testing for `errors` json files even when `n_polluted` = 0 (no pollution, i.e. no error.)

**Comment by chainsawriot** [comment]
@TimBMK Question: When the parameter `errors` is true and there is no error, `get_tweets` will generate an empty `errors_****.json` file. Is it the intended behavior? 
I am asking this because it is currently etched in the test case. If it is not the intended behavior, I'll need to change the test case.

https://github.com/cjbarrie/academictwitteR/pull/264/files#r772005557

**Comment by TimBMK** [comment]
Good question, I hadn't encountered that case. Keeping them makes for a slightly tidier data structure (I'd say), but will potentially create heaps of empty data files. What do you think? Maybe add a check in df_to_json to prevent empty error files being written?

**Comment by chainsawriot** [comment]
@TimBMK I actually don't mind having all those empty json files to maintain a trio structure (`data`, `users`, and `errors`) of json files of each query. So I think I'll keep them.

**Comment by chainsawriot** [comment]
@TimBMK OK, it's there. Please note that the output is slightly different.

``` r
devtools::load_all("~/dev/academictwitteR")
#> ℹ Loading academictwitteR
fff <- readRDS("~/dev/academictwitteR/tests/testdata/fff_de.RDS")
polluted_fff <- c("1", "2", "3", fff, "4")
emptydir <- academictwitteR:::.gen_random_dir()  
res <- hydrate_tweets(polluted_fff, verbose = TRUE, errors = TRUE, bind_tweets = TRUE, data_path = emptydir)
#> Batch 1 out of 2 : ids 1 to 1266817176687779846 
#> Total  3  tweet(s) can't be retrieved.
#> Total of  97  out of  113  tweet(s) retrieved.
#> Batch 2 out of 2 : ids 1266816778010845196 to 4 
#> Total  4  tweet(s) can't be retrieved.
#> Total of  109  out of  113  tweet(s) retrieved.
unlink(emptydir, recursive = TRUE)
```

<sup>Created on 2021-12-21 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

@cjbarrie @justinchuntingho and @TimBMK 

Please note that I've added >400 test cases (> 50% of the tests) to vigorously test the helper function `get_tweets` with the error capturing mechanism. I think it is warranted because the cash cow function `get_all_tweets` (and many wrappers around it) is depending on the super stable `get_tweets` helper function. @TimBMK Please don't interpret that as I don't trust your code. I just want to make sure things work as intended.

This is quite a big change. As the AUTs, @cjbarrie and @justinchuntingho I suggest you guys give it a test; review if needed. I think I will pass it back to you from now on. Please "Squash and Merge".

Thank you very much.

**Comment by TimBMK** [comment]
Great work! I'm happy you're doing all the testing. The last thing I want is to break things with new functionality. I've proposed a small change for the verbose output of hydrate_tweets, so it no longer reports more tweets as retrieved than were originally queued (the tweet endpoint does not allow more than 100 ids, so the maximum would be 100 - I suspect the additional 13 in your example are a product of the error messages?)

**Comment by chainsawriot** [comment]
@TimBMK the length of `fff` is 109 and then it is polluted with 4 junks, so that's 113. In the nutshell, as the batch size is always 100, I think it makes more sense to display the # of errors rather than the # of returned tweets. After the collection of each batch, it just displays how long it is from the goal.

**Comment by TimBMK** [comment]
@justinchuntingho  Ah that makes more sense, now I get it. You can ignore my commit, then. It's a bit different from what I originally intended, but your solution may actually be more useful for the user. We could consider moving the verbose output of totals to after the loop, rather than displaying it for every batch, even. But I'll leave that decision to you

**Comment by chainsawriot** [comment]
@cjbarrie If you really want to do this, I rather do it this way:

```r
resume_hydration <- function(ids, data_path, error = FALSE, verbose = TRUE) {
  if(!dir.exists(file.path(data_path))){
    stop("Directory ", data_path, " doesn't exist.")
  }
  existing_df <- bind_tweets(data_path, verbose = verbose)
  uncollected_ids <- setdiff(ids, existing_df$id)
  hydrate_tweets(uncollected_ids, data_path = data_path, error = error, verbose = verbose)
}
```

But once again, keep it small. One feature, one PR into one commit, at a time. This resume would need another trainload of tests.

**Comment by cjbarrie** [comment]
Yeah, sure. But we could easily avoid the previously errored tweet IDs by not doing as you suggest but instead something like:

```r

resume_hydration <- function(ids, data_path, error = FALSE, verbose = TRUE) {
  if(!dir.exists(file.path(data_path))){
    stop("Directory ", data_path, " doesn't exist.")
  }
  existing_df <- bind_tweets(data_path, verbose = verbose)
  lastid <- tail(existing_df, n =1)$id
  pos <- match(lastid,ids)
  uncollected_ids <- ids[pos:length(ids)]
  hydrate_tweets(uncollected_ids, data_path = data_path, error = error, verbose = verbose)
}

```

Anyway, will move this to separate item TBC. Happy to merge but will wait a bit to see if @justinchuntingho has anything to add

**Comment by cjbarrie** [comment]
I'm happy to merge this now unless you would prefer waiting for @justinchuntingho to review as well

**Comment by chainsawriot** [comment]
@cjbarrie Your call.

**Comment by cjbarrie** [comment]
I'm happy with how things look. Thank you very much both for this contribution

**Pull request merged**